diff --git a/ggml/src/ggml-sycl/vecdotq.hpp b/ggml/src/ggml-sycl/vecdotq.hpp
index 765fb7f15..c34e6a4fe 100644
--- a/ggml/src/ggml-sycl/vecdotq.hpp
+++ b/ggml/src/ggml-sycl/vecdotq.hpp
@@ -800,7 +800,9 @@ vec_dot_q5_1_q8_1_impl(const int *vl, const int *vh, const int *u,
     return sumi*d5d8 + m5s8 / (QI5_1 / vdr);
 }
 
+#ifndef VDR_Q8_0_Q8_1_MMVQ
 #define VDR_Q8_0_Q8_1_MMVQ 2
+#endif
 #define VDR_Q8_0_Q8_1_MMQ 8
 
 template <int vdr>
