diff --git a/csrc/xpu/gdn_attn/spec_decode.hpp b/csrc/xpu/gdn_attn/spec_decode.hpp
index 4b36044..3597048 100644
--- a/csrc/xpu/gdn_attn/spec_decode.hpp
+++ b/csrc/xpu/gdn_attn/spec_decode.hpp
@@ -584,7 +584,6 @@ struct store_conv_states_kernel {
       const int total_spec_tokens,
       const int num_actual_tokens,
       const int state_cols,
-      const int state_len,
       const int num_k_heads,
       const int head_k_dim,
       const int num_v_heads,
